### PR TITLE
feat: Add `IFilenameValidator` to have one consistent place for filename validation

### PR DESCRIPTION
**Pull request by susnux**
* Part of #44963 

## Summary

This is the first part, split https://github.com/nextcloud/server/pull/45151 to make review easier.

This adds a new interface `IFilenameValidator` that allows validating filenames to have one consistent location for filename handling.
We need two¹ functions:
1. Validation with meaningful messages (as errors)
2. Validation for `if` with boolean return

¹ We need in some rare private filesystem locations a third function: We need to check if a file is forbidden but can not validate all other restrictions, because we use our filesystem API also for setup where we use possibly forbidden characters as we are only setup Nextcloud itself not files later synced.

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [x] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Files/FilenameValidator.php [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	public function getForbiddenCharacters(): array {
+		if (empty($this->forbiddenCharacters)) {
+			// Get always forbidden characters
+			$forbiddenCharacters = str_split(\OCP\Constants::FILENAME_INVALID_CHARS);
+			if ($forbiddenCharacters === false) {
+				$forbiddenCharacters = [];
+			}
+
+			// Get admin defined invalid characters
+			$additionalChars = $this->config->getSystemValue('forbidden_chars', []);
```
Comment (come-nc) [inline lib/private/Files/FilenameValidator.php:126]:
```suggestion
			$additionalChars = $this->config->getSystemValue('forbidden_filename_characters', []);
```
You used the legacy key twice I think?

File: lib/private/Files/FilenameValidator.php [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+			throw new FileNameTooLongException();
+		}
+
+		if ($this->isForbidden($filename)) {
+			throw new ReservedWordException();
+		}
+
+		$this->checkForbiddenExtension($filename);
+
+		$this->checkForbiddenCharacters($filename);
```
Comment (come-nc) [inline lib/private/Files/FilenameValidator.php:188]: Shouldn’t all these tests be made on the trimmed var instead?

File: lib/private/Files/FilenameValidator.php [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	}
+
+	/**
+	 * Check if the filename is forbidden
+	 * @param string $filename
```
Comment (come-nc) [inline lib/private/Files/FilenameValidator.php:193]: Can be a path if I understand correctly, as basename is called before anything else.

**Comment by susnux** [comment]
Follow up with usage here: https://github.com/nextcloud/server/pull/46375

File: lib/private/Files/FilenameValidator.php (hunk 2 above)
Comment (susnux) [reply lib/private/Files/FilenameValidator.php:188]: No having whitespace in the name is pretty valid, so rules might apply to those whitespace names.
The trim is only required for the empty check

File: lib/private/Files/FilenameValidator.php (hunk 3 above)
Comment (susnux) [reply lib/private/Files/FilenameValidator.php:193]: Yes sadly required for some function calls in very internal filesystem logic :pensive: 
I will rename it

Comment (susnux) [reply lib/private/Files/FilenameValidator.php]: done

File: lib/private/Files/FilenameValidator.php [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+			// Get always forbidden characters
+			$forbiddenCharacters = str_split(\OCP\Constants::FILENAME_INVALID_CHARS);
+			if ($forbiddenCharacters === false) {
+				$forbiddenCharacters = [];
+			}
+
+			// Get admin defined invalid characters
+			$additionalChars = $this->config->getSystemValue('forbidden_filename_characters', []);
+			if (!is_array($additionalChars)) {
+				$this->logger->error('Invalid system config value for "forbidden_chars" is ignored.');
```
Comment (come-nc) [inline lib/private/Files/FilenameValidator.php:128]:
```suggestion
				$this->logger->error('Invalid system config value for "forbidden_filename_characters" is ignored.');
```

**Review by come-nc — COMMENTED** [review]
Almost there

File: lib/private/Files/FilenameValidator.php (hunk 4 above)
Comment (susnux) [reply lib/private/Files/FilenameValidator.php]: :see_no_evil:

**Review by come-nc — APPROVED** [review]

**Review by sorbaugh — APPROVED** [review]

**Pull request merged**